src/modules/atomic-queue/AtomicQueue.sol

```diff
@@ -0,0 +1,298 @@
+    error AtomicQueue__RequestDeadlineExceeded(address user);
+    error AtomicQueue__UserNotInSolve(address user);
+    error AtomicQueue__ZeroOfferAmount(address user);
+
+    //============================== EVENTS ===============================
+
+    /**
+     * @notice Emitted when `updateAtomicRequest` is called.
+     */
+    event AtomicRequestUpdated(
+        address user,
+        address offer,
+        address want,
+        uint256 amount,
+        uint256 deadline,
+        uint256 minPrice,
+        uint256 timestamp
+    );
+
+    /**
+     * @notice Emitted when `solve` exchanges a users shares for the underlying asset.
+     */
+    event AtomicRequestFulfilled(
+        address user,
+        address offer,
+        address want,
+        uint256 sharesSpent,
+        uint256 assetsReceived,
+        uint256 timestamp
+    );
+
+    //============================== USER FUNCTIONS ===============================
+
+    /**
+     * @notice Get a users Withdraw Request.
+     */
+    function getUserAtomicRequest(address user, ERC20 offer, ERC20 want) external view returns (AtomicRequest memory) {
+        return userAtomicRequest[user][offer][want];
+    }
+
+    /**
+     * @notice Helper function that returns either
+     *         true: Withdraw request is valid.
+     *         false: Withdraw request is not valid.
+     * @dev It is possible for a withdraw request to return false from this function, but using the
+     *      request in `updateAtomicRequest` will succeed, but solvers will not be able to include
+     *      the user in `solve` unless some other state is changed.
+     */
+    function isAtomicRequestValid(
+        ERC20 offer,
+        address user,
+        AtomicRequest calldata userRequest
+    ) external view returns (bool) {
+        // Validate amount.
+        if (userRequest.offerAmount > offer.balanceOf(user)) return false;
+        // Validate deadline.
+        if (block.timestamp > userRequest.deadline) return false;
+        // Validate approval.
+        if (offer.allowance(user, address(this)) < userRequest.offerAmount) return false;
+        // Validate offerAmount is nonzero.
+        if (userRequest.offerAmount == 0) return false;
+        // Validate atomicPrice is nonzero.
+        if (userRequest.atomicPrice == 0) return false;
+
+        return true;
+    }
+
+    /**
+     * @notice Allows user to add/update their withdraw request.
+     * @notice It is possible for a withdraw request with a zero atomicPrice to be made, and solved.
+     *         If this happens, users will be selling their shares for no assets in return.
+     *         To determine a safe atomicPrice, share.previewRedeem should be used to get
+     *         a good share price, then the user can lower it from there to make their request fill faster.
+     */
+    function updateAtomicRequest(ERC20 offer, ERC20 want, AtomicRequest calldata userRequest) external nonReentrant {
```

Comment:
The current withdraw queue uses 51k gas on a user deposit, which is pretty poggers. If we add in validation checks, increase gas usage by 20k gas or so, which is pretty significant given the context. I still think its best to not validate, since of all the things to validate the one that really matters is the atomic price, but as mentioned in another comment, doing that is not really worth the added baggage. So then we are just increasing the gas cost by 40%, and adding very little benefit to users that interact with the queue through our UI(which will be the vast majority of users)

Reply:
Thanks for the breakdown. I agree with your take on this now that you have outlined it out in detail. Nice! 😊